corehq/apps/locations/bulk_management.py

```diff
@@ -0,0 +1,131 @@
+"""
+Bulk rearrange locations.
+
+This includes support for changing location types, changing locations' parents,
+deleting things, and so on.  See the spec doc for specifics:
+https://docs.google.com/document/d/1gZFPP8yXjPazaJDP9EmFORi88R-jSytH6TTgMxTGQSk/
+"""
+from collections import namedtuple, Counter, defaultdict
+
+from dimagi.utils.decorators.memoized import memoized
+
+from .tree_utils import BadParentError, CycleError, assert_no_cycles
+
+LocationTypeStub = namedtuple(
+    "LocationTypeStub",
+    "name code parent_code shares_cases view_descendants"
+)
+
+LocationStub = namedtuple(
+    "LocationStub",
+    "name site_code location_type parent_code location_id external_id latitude longitude"
+)
+
+
+class LocationTreeValidator(object):
+    def __init__(self, location_types, locations):
+        self.location_types = location_types
+        self.locations = locations
```

Comment:
`namedtuple`s - I wrote this to test the internal consistency of a snapshot of types and locations.  The idea is that you can run this against an upload to make sure everything makes sense before applying any changes or even touching the DB.  If they say "I want my organization to look like this", you can verify that - for instance - all codes are unique, and all locations have appropriate types and parents before attempting to apply any of the changes.